File: src/pasta/shamir.rs
//! Ed25519 のスカラー体上の Shamir 秘密分散。
//!
//! 既存の `crate::field` は GF(2^61-1) 上で動いているが、閾値署名と TOPRF は
//! 曲線のスカラー体 (ℓ ≈ 2^252) 上でなければならないので、こちらを別に用意する。

use curve25519_dalek::Scalar;
use rand::prelude::*;

pub type ParticipantId = u16;

#[derive(Clone, Copy, Debug)]
pub struct Share {
    pub id: ParticipantId,
    pub value: Scalar,
}

/// OS 乱数 64 バイトを法 ℓ に落として一様なスカラーを得る。
pub fn random_scalar() -> Scalar {
    let mut bytes = [0u8; 64];
    rand::rng().fill_bytes(&mut bytes);
    Scalar::from_bytes_mod_order_wide(&bytes)
}

/// `secret` を (t, n) 閾値で分散する。x 座標は 1..=n。
pub fn split(secret: &Scalar, n: ParticipantId, t: ParticipantId) -> Vec<Share> {
    assert!(t >= 1 && t <= n, "1 <= t <= n が必要");

    // f(x) = secret + c_1 x + ... + c_{t-1} x^{t-1}
    let coeffs: Vec<Scalar> = (1..t).map(|_| random_scalar()).collect();

    (1..=n)
        .map(|id| {
            let x = Scalar::from(id as u64);
            // Horner 法で f(x) を評価する
            let mut acc = Scalar::ZERO;
            for c in coeffs.iter().rev() {
                acc = acc * x + c;
            }
            Share {
                id,
                value: acc * x + secret,
            }
        })
        .collect()
}

/// 参加者集合 `ids` における `target` のラグランジュ係数 λ_i = Π_{j≠i} x_j / (x_j - x_i)。
pub fn lagrange_coeff(ids: &[ParticipantId], target: ParticipantId) -> Scalar {
    let xi = Scalar::from(target as u64);
    let mut num = Scalar::ONE;
    let mut den = Scalar::ONE;

    for &j in ids {
        if j == target {
            continue;
        }
        let xj = Scalar::from(j as u64);
        num *= xj;
        den *= xj - xi;
    }

    num * den.invert()
}

/// t 個以上のシェアから秘密を復元する。テストと、鍵が「揃っていない」ことの確認用。
pub fn reconstruct(shares: &[Share]) -> Scalar {
    let ids: Vec<ParticipantId> = shares.iter().map(|s| s.id).collect();
    shares
        .iter()
        .map(|s| lagrange_coeff(&ids, s.id) * s.value)
        .fold(Scalar::ZERO, |acc, v| acc + v)
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn t_of_n_reconstructs() {
        let secret = random_scalar();
        let shares = split(&secret, 5, 3);

        // 任意の 3 枚で復元できる
        assert_eq!(reconstruct(&shares[0..3]), secret);
        assert_eq!(reconstruct(&[shares[0], shares[2], shares[4]]), secret);
    }

    #[test]
    fn fewer_than_t_reveals_nothing() {
        let secret = random_scalar();
        let shares = split(&secret, 5, 3);

        // 2 枚だけでは復元できない（ラグランジュ補間の次数が足りない）
        assert_ne!(reconstruct(&shares[0..2]), secret);
    }
}

File: src/crypto/shamir.ts
import { ed25519 } from '@noble/curves/ed25519';

/**
 * The order of the Ed25519 prime-order subgroup / scalar field:
 * ℓ = 2^252 + 27742317777372353535851937790883648493
 */
export const ED25519_ORDER: bigint = ed25519.CURVE.n;

/**
 * Participant identifier (positive integer, e.g. 1, 2, ...).
 */
export type ParticipantId = number;

/**
 * A secret share evaluated at participant's ID.
 */
export interface Share {
  id: ParticipantId;
  value: bigint;
}

/**
 * Standard positive modulo arithmetic.
 */
export function mod(a: bigint, m: bigint = ED25519_ORDER): bigint {
  const r = a % m;
  return r >= 0n ? r : r + m;
}

/**
 * Modular inverse using Fermat's Little Theorem: a^(p-2) mod p.
 */
export function invert(a: bigint, m: bigint = ED25519_ORDER): bigint {
  const norm = mod(a, m);
  if (norm === 0n) {
    throw new Error('Zero has no modular inverse');
  }
  return modPow(norm, m - 2n, m);
}

/**
 * Modular exponentiation: (base^exp) mod m.
 */
export function modPow(base: bigint, exp: bigint, m: bigint): bigint {
  let res = 1n;
  let b = mod(base, m);
  let e = exp;
  while (e > 0n) {
    if ((e & 1n) === 1n) {
      res = mod(res * b, m);
    }
    b = mod(b * b, m);
    e >>= 1n;
  }
  return res;
}

/**
 * Converts little-endian bytes to a BigInt.
 */
export function bytesToBigIntLE(bytes: Uint8Array): bigint {
  let acc = 0n;
  for (let i = bytes.length - 1; i >= 0; i--) {
    acc = (acc << 8n) | BigInt(bytes[i]);
  }
  return acc;
}

/**
 * Converts a BigInt to a little-endian Uint8Array of fixed length (default 32 bytes).
 */
export function bigIntToBytesLE(value: bigint, length: number = 32): Uint8Array {
  const norm = mod(value, ED25519_ORDER);
  const out = new Uint8Array(length);
  let cur = norm;
  for (let i = 0; i < length; i++) {
    out[i] = Number(cur & 0xffn);
    cur >>= 8n;
  }
  return out;
}

/**
 * Converts 32-byte canonical little-endian representation to scalar mod ℓ.
 */
export function bytesToScalar(bytes: Uint8Array): bigint {
  return mod(bytesToBigIntLE(bytes), ED25519_ORDER);
}

/**
 * Converts scalar to 32-byte canonical little-endian representation.
 */
export function scalarToBytes(scalar: bigint): Uint8Array {
  return bigIntToBytesLE(scalar, 32);
}

/**
 * Generates a cryptographically uniform random scalar mod ℓ by taking 64 bytes
 * of OS entropy and reducing mod ℓ (equivalent to Scalar::from_bytes_mod_order_wide).
 */
export function randomScalar(): bigint {
  const bytes = new Uint8Array(64);
  globalThis.crypto.getRandomValues(bytes);
  return mod(bytesToBigIntLE(bytes), ED25519_ORDER);
}

/**
 * Splits a secret scalar into n shares with threshold t using Shamir's secret sharing scheme.
 * The polynomial is f(x) = secret + c_1*x + ... + c_{t-1}*x^{t-1} mod ℓ.
 * Participant IDs are 1-indexed (1, 2, ..., total).
 *
 * @param secret The secret scalar to share (f(0) = secret).
 * @param threshold The minimum number of shares required to reconstruct the secret (t).
 * @param total The total number of shares to produce (n).
 */
export function splitSecret(secret: bigint, threshold: number, total: number): Share[] {
  if (threshold < 1 || threshold > total) {
    throw new Error(`Invalid threshold: must satisfy 1 <= threshold (${threshold}) <= total (${total})`);
  }

  // Generate random coefficients c_1, ..., c_{t-1}
  const coeffs: bigint[] = [];
  for (let i = 1; i < threshold; i++) {
    coeffs.push(randomScalar());
  }

  const shares: Share[] = [];
  for (let id = 1; id <= total; id++) {
    const x = BigInt(id);
    // Evaluate f(x) using Horner's method
    let acc = 0n;
    for (let j = coeffs.length - 1; j >= 0; j--) {
      acc = mod(mod(acc * x, ED25519_ORDER) + coeffs[j], ED25519_ORDER);
    }
    const val = mod(mod(acc * x, ED25519_ORDER) + secret, ED25519_ORDER);
    shares.push({ id, value: val });
  }

  return shares;
}

/**
 * Computes the Lagrange basis polynomial coefficient λ_i(0) for participant `targetId`
 * evaluated at 0 with respect to the set of participant `ids`:
 * λ_i(0) = ∏_{j ∈ ids, j ≠ i} (x_j / (x_j - x_i)) mod ℓ.
 */
export function lagrangeCoeff(ids: number[], targetId: number): bigint {
  const xi = BigInt(targetId);
  let num = 1n;
  let den = 1n;

  for (const j of ids) {
    if (j === targetId) continue;
    const xj = BigInt(j);
    num = mod(num * xj, ED25519_ORDER);
    den = mod(den * (xj - xi), ED25519_ORDER);
  }

  return mod(num * invert(den, ED25519_ORDER), ED25519_ORDER);
}

/**
 * Combines t or more shares using Lagrange interpolation to reconstruct the secret scalar.
 * Reconstructed secret = ∑_{i} λ_i(0) * y_i mod ℓ.
 *
 * @param shares Array of shares from distinct participants.
 */
export function combineShares(shares: Share[]): bigint {
  if (shares.length === 0) {
    throw new Error('At least one share is required to combine');
  }

  const ids = shares.map((s) => s.id);
  const uniqueIds = new Set(ids);
  if (uniqueIds.size !== ids.length) {
    throw new Error('Duplicate participant IDs found in shares');
  }

  let secret = 0n;
  for (const s of shares) {
    const lambda = lagrangeCoeff(ids, s.id);
    secret = mod(secret + mod(lambda * s.value, ED25519_ORDER), ED25519_ORDER);
  }

  return secret;
}
